rest/data/host_create.go

```diff
@@ -0,0 +1,228 @@
+package data
+
+import (
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/evergreen-ci/evergreen"
+	"github.com/evergreen-ci/evergreen/apimodels"
+	"github.com/evergreen-ci/evergreen/cloud"
+	"github.com/evergreen-ci/evergreen/command"
+	"github.com/evergreen-ci/evergreen/model"
+	"github.com/evergreen-ci/evergreen/model/distro"
+	"github.com/evergreen-ci/evergreen/model/host"
+	"github.com/evergreen-ci/evergreen/model/task"
+	"github.com/evergreen-ci/evergreen/model/user"
+	"github.com/evergreen-ci/gimlet"
+	"github.com/mitchellh/mapstructure"
+	"github.com/mongodb/grip"
+	"github.com/pkg/errors"
+)
+
+// DBCreateHostConnector supports `host.create` commands from the agent.
+type DBCreateHostConnector struct{}
+
+// ListHostsForTask lists running hosts scoped to the task or the task's build.
+func (dc *DBCreateHostConnector) ListHostsForTask(taskID string) ([]host.Host, error) {
+	t, err := task.FindOneId(taskID)
+	if err != nil {
+		return nil, gimlet.ErrorResponse{StatusCode: http.StatusInternalServerError, Message: "error finding task"}
+	}
+	if t == nil {
+		return nil, gimlet.ErrorResponse{StatusCode: http.StatusInternalServerError, Message: "no task found"}
+	}
+
+	catcher := grip.NewBasicCatcher()
+	hostsSpawnedByTask, err := host.FindHostsSpawnedByTask(t.Id)
+	catcher.Add(err)
+	hostsSpawnedByBuild, err := host.FindHostsSpawnedByBuild(t.BuildId)
+	catcher.Add(err)
+	if catcher.HasErrors() {
+		return nil, gimlet.ErrorResponse{StatusCode: http.StatusInternalServerError, Message: catcher.String()}
+	}
+	hosts := []host.Host{}
+	for _, h := range hostsSpawnedByBuild {
+		hosts = append(hosts, h)
+	}
+	for _, h := range hostsSpawnedByTask {
+		hosts = append(hosts, h)
+	}
+	return hosts, nil
+}
+
+func (dc *DBCreateHostConnector) CreateHostsFromTask(t *task.Task, user user.DBUser, keyNameOrVal string) error {
+	if t == nil {
+		return errors.New("no task to create hosts from")
+	}
+
+	keyVal, err := user.GetPublicKey(keyNameOrVal)
+	if err != nil {
+		keyVal = keyNameOrVal
+	}
+
+	tc, err := model.MakeConfigFromTask(t)
+	if err != nil {
+		return err
+	}
+
+	projectTask := tc.Project.FindProjectTask(tc.Task.DisplayName)
+	if projectTask == nil {
+		return errors.Errorf("unable to find configuration for task %s", tc.Task.Id)
+	}
+
+	createHostCmds := []*command.CreateHost{}
+	catcher := grip.NewBasicCatcher()
+	for _, commandConf := range projectTask.Commands {
+		cmds, err := command.Render(commandConf, tc.Project.Functions)
+		if err != nil {
+			return errors.Wrap(err, "unable to parse command")
+		}
+		for _, cmd := range cmds {
+			createHost, ok := cmd.(*command.CreateHost)
```

Comment:
you can filter on cmd.Name() rather than type casting.

Reply:
I need to access one of the command's properties below though. I don't see a way to do this without type casting, though if you have suggestions I'm open to them